pallet/src/migrations/v0_to_v1.rs

```diff
@@ -22,7 +22,14 @@ pub fn migrate<T: Config, P: GetStorageVersion + PalletInfoAccess>() -> Weight {
         log::info!(target: "pallet_aleph", "Running migration from STORAGE_VERSION 0 to 1");
 
         let mut writes = 0;
-        match SessionForValidatorsChange::translate(sp_std::convert::identity) {
+        match SessionForValidatorsChange::translate(|old: Option<Option<u32>>| {
+            log::info!(target: "pallet_aleph", "Current storage value for SessionForValidatorsChange {:?}", old);
+            match old {
+                Some(None) | Some(Some(0u32)) => None,
+                Some(value) => Some(value),
```

Comment:
This returns `Option<Option<..>>`. Don't you want to `flatten` it (modulo the 0 thing, which is kinda annoying)?

Reply:
it *was* a double wrapped type, now it's a single option over the value. Proof in the pudding (logs from devnet where migration was applied):

```
2-09 14:05:44 Current storage value for SessionForValidatorsChange Some(None)                                                                                                                                                                                            
2021-12-09 14:05:44 Succesfully migrated storage for SessionForValidatorsChange     
```

```
2021-12-09 14:09:16 @new_session 63533                                                                                                                                                                                                                                         
2021-12-09 14:09:16 @new_session Pallet::<T>::session_for_validators_change() None                                                                                                                                                                                             
2021-12-09 14:09:16 ✨ Imported #317660 (0xdbb4…7d02)                                                                                                                                                                                                                          
2021-12-09 14:09:17 💤 Idle (9 peers), best: #317660 (0xdbb4…7d02), finalized #317659 (0x7509…1f6d), ⬇ 5.9kiB/s ⬆ 4.5kiB/s 
```

Reply:
I think I know what you mean, but it's a bit cumbersome to express using this translate mapping and what is stored atm:
```
        match SessionForValidatorsChange::translate(|old: Option<_>| {
            log::info!(target: "pallet_aleph", "Current storage value for SessionForValidatorsChange {:?}", old);
            match old {
                Some(None) | Some(Some(0u32)) | None => None,
                _ => old,
            }
        })
```

Reply:
I really don't get how this works then – does Some(Some(X)) correctly deserialize as Some(X) and you are deliberately using that fact? Because formally you are returning `Option<Option<...>>` from this code and this is what bothers me. <_<

Reply:
Something like this yes, it's a bit cumbersome, i know :/

We had a ValueQuery over an Option<T> type, with a default for T None, which is how we ended up with sth that decodes to Some(None) and effectively is of type Option<Option<T>>:

```
    #[pallet::type_value]
    pub fn DefaultSessionForValidatorsChange<T: Config>() -> Option<u32> {
        None
    }

    #[pallet::storage]
    #[pallet::getter(fn session_for_validators_change)]
    pub type SessionForValidatorsChange<T: Config> =
        StorageValue<_, Option<u32>, ValueQuery, DefaultSessionForValidatorsChange<T>>;
```

Now translate operates from and to Option<_>, so there is no way to double unpack and return, won't compile AFAIK, I tried. It was designed for cases like i32 -> u32.

But eventually we want to end up storing sth that decodes to an Option<T>, that the storage layer can query with OptionQuery over vanilla T type, so we want to store sth that:

- If there was a value (we are in a middle of change) decodes to Option<T>
- else decodes to None

Given that we control when we apply the migration, and can be sure now validator change was scheduled, just removing the storage item, or overwriting it with None would also work - storage layer evicts None. But to be super correct this is covered as well.

Reply:
@timorl I completely mixed the semantics of translate and @DamianStraszak pointed it out: it takes the *new* storage type and a function that maps it from the old to new value. With this it's possible, to go from Option<Option> to Option.

Reply:
After the changes in `translate` everything makes sense to me when it comes to types. @timorl you can also take another look now